README.md
# mdsanima-dev

[![license-mdsanima][badge-01]][link-01]
[![latest-version-on-pypi][badge-02]][link-02]
[![github-top-language][badge-03]][link-01]
[![pypi-python-version][badge-04]][link-02]
[![code-style-black][badge-05]][link-03]

[![pype-total-downloads][badge-pe]][link-04]
[![github-repo-stars][badge-06]][link-01]
[![github-open-issues][badge-07]][link-05]
[![github-closed-issues][badge-08]][link-06]
[![github-deployments][badge-09]][link-07]
[![github-pages-documentation][badge-10]][link-08]

_Python_ package [mdsanima-dev][link-02] is for colorizing, adding emoji and making table on console shell print output.
You can find here _Python_ modules that's make coding easier and cooler, includes several useful modules that we're use
in various projects. In the future we're add more cool function on this package.

Docomentation available at [GitHub Pages][link-08], be sure to check it.

## Installation `Options`

Instructions how to install the _Python_ package [mdsanima-dev][link-02] on your system.

<!-- start help-installation -->

Options [PyPI](#using-pypi) allows you to install the package **globally** on your system. If you want to install the
package in an **isolated environment** on your system use the [virtualenv](#using-virtualenv) options.

<!-- end help-installation -->

### Using `PyPI`

<!-- start help-using-pypi -->

Install latest version:

```shell
python3 -m pip install mdsanima-dev
```

Install specific version:

```shell
python3 -m pip install mdsanima-dev==0.2.0
```

Upgrade package to latest version:

```shell
python3 -m pip install --upgrade mdsanima-dev
```

Reinstall package to latest version:

```shell
python3 -m pip install --force-reinstall msdsanima-dev
```

Uninstall package:

```shell
python3 -m pip uninstall mdsanima-dev
```

<!-- end help-using-pypi -->

### Using `virtualenv`

<!-- start help-on-virtualenv -->

Clone repository with all number of commits history using HTTPS:

```shell
git clone https://github.com/mdsanima-dev/mdsanima-dev.git
```

Clone repository with the specified number of commits history using SSH:

```shell
git clone --depth=1 git@github.com:mdsanima-dev/mdsanima-dev.git
```

Then go to the repository directory:

```shell
cd mdsanima-dev
```

Create isolated environment with specific package version then activate and install requirements, type in terminal:

```shell
virtualenv --setuptools 67.7.2 --wheel 0.40.0 --pip 23.1.2 .venv
source .venv/bin/activate
pip install -r requirements-dev.txt
pip install -r requirements-doc.txt
```

Install build tools for creating distribution, then build package from source, type in terminal:

```shell
pip install --upgrade build
python -m build
```

Finnaly install the package in virtual environment, type in the terminal:

```shell
pip install --force-reinstall dist/mdsanima_dev-0.2.0-py3-none-any.whl
```

In this case I used the `--force-reinstall` option because I had an earlier version of this package already installed.

Also you can build Sphinx documentation, type in terminal:

```shell
python setup.py build_sphinx
cd build/dirhtml
python -m http.server 8080
```

Now open [localhost](http://localhost:8080) to see the documentation.

Deactivate virtual environment, type in the terminal:

```shell
deactivate
```

<!-- end help-on-virtualenv -->

## Develempent

Instruction for the _Python_ package development. Please check the [documentation][link-08] site for more information.

## Showcase

![mdsanima-dev-python-show](docs/_images/gif/mdsanima_dev_python_show.gif)

## Follow Me

These are my social media account, be sure to check it. Thanks!

[![github-followers-mdsanima][badge-11]][link-10]
[![twitter-follow-toudajew][badge-12]][link-12]
[![twitter-follow-str9led][badge-13]][link-13]
[![twitter-follow-mdsanima][badge-14]][link-14]

[![subreddit-subscribers-mdsanima][badge-15]][link-15]
[![youtube-subscribers-mdsanima][badge-16]][link-16]
[![youtube-views-mdsanima][badge-17]][link-16]
[![twitch-status-mdsanima][badge-18]][link-17]
[![discord-chat-mdsanima][badge-19]][link-09]

## License

Python package [mdsanima-dev][link-02] developed by [Marcin Różewski][link-10] is released under the terms of
[MIT License][link-11]

[badge-pe]: https://static.pepy.tech/personalized-badge/mdsanima-dev?period=total&units=none&left_color=grey&right_color=yellowgreen&left_text=downloads
[badge-01]: https://img.shields.io/github/license/mdsanima-dev/mdsanima-dev?style=flat
[badge-02]: https://img.shields.io/pypi/v/mdsanima-dev?style=flat&logo=pypi&logoColor=lightgray
[badge-03]: https://img.shields.io/github/languages/top/mdsanima-dev/mdsanima-dev?style=flat&logo=python&logoColor=lightgray
[badge-04]: https://img.shields.io/pypi/pyversions/mdsanima-dev?style=flat&logo=python&logoColor=lightgray
[badge-05]: https://img.shields.io/badge/code%20style-black-000000.svg?logo=python&logoColor=lightgray
[badge-06]: https://img.shields.io/github/stars/mdsanima-dev/mdsanima-dev?style=flat&logo=github
[badge-07]: https://img.shields.io/github/issues-raw/mdsanima-dev/mdsanima-dev?style=flat&logo=github
[badge-08]: https://img.shields.io/github/issues-closed-raw/mdsanima-dev/mdsanima-dev?style=flat&logo=github
[badge-09]: https://img.shields.io/github/deployments/mdsanima-dev/mdsanima-dev/github-pages?style=flat&logo=github
[badge-10]: https://img.shields.io/website?url=https%3A%2F%2Fmdsanima-dev.github.io%2Fmdsanima-dev%2F?style=flat&logo=github
[badge-11]: https://img.shields.io/github/followers/mdsanima?style=social
[badge-12]: https://img.shields.io/twitter/follow/toudajew?style=social
[badge-13]: https://img.shields.io/twitter/follow/str9led?style=social
[badge-14]: https://img.shields.io/twitter/follow/mdsanima?style=social
[badge-15]: https://img.shields.io/reddit/subreddit-subscribers/mdsanima?style=social
[badge-16]: https://img.shields.io/youtube/channel/subscribers/UCB5na2BRwrnwx00LCspbG5Q?style=social
[badge-17]: https://img.shields.io/youtube/channel/views/UCB5na2BRwrnwx00LCspbG5Q?style=social
[badge-18]: https://img.shields.io/twitch/status/mdsanima?style=social
[badge-19]: https://img.shields.io/discord/621477380359454742?style=social&logo=discord
[link-01]: https://github.com/mdsanima-dev/mdsanima-dev
[link-02]: https://pypi.org/project/mdsanima-dev
[link-03]: https://github.com/psf/black
[link-04]: https://pepy.tech/project/mdsanima-dev
[link-05]: https://github.com/mdsanima-dev/mdsanima-dev/issues?q=is%3Aopen+is%3Aissue
[link-06]: https://github.com/mdsanima-dev/mdsanima-dev/issues?q=is%3Aissue+is%3Aclosed
[link-07]: https://github.com/mdsanima-dev/mdsanima-dev/deployments/activity_log?environment=github-pages
[link-08]: https://mdsanima-dev.github.io/mdsanima-dev/
[link-09]: https://discord.gg/c3m7pTF
[link-10]: https://github.com/mdsanima
[link-11]: https://github.com/mdsanima-dev/mdsanima-dev/blob/master/LICENSE
[link-12]: https://twitter.com/intent/follow?toudajew&screen_name=toudajew
[link-13]: https://twitter.com/intent/follow?str9led&screen_name=str9led
[link-14]: https://twitter.com/intent/follow?mdsanima&screen_name=mdsanima
[link-15]: https://reddit.com/r/mdsanima/
[link-16]: https://youtube.com/mdsanima?sub_confirmation=1
[link-17]: https://twitch.tv/mdsanima/


CHANGELOG.md
# {octicon}`checklist;1em;sd-text-secondary` `CHANGELOG`

All notable changes to this project will be documented in this file.

Please check
[standard-version {octicon}`link-external;0.8em`](https://github.com/conventional-changelog/standard-version)
and documentation for commit guidelines. Also you should check
[conventional-changelog {octicon}`link-external;0.8em`](https://github.com/mdsanima/conventional-changelog)
for more detail.

Docomentation for _Python_ package
[mdsanima-dev {octicon}`link-external;0.8em`](https://pypi.org/project/mdsanima-dev) is available
at [GitHub Pages {octicon}`link-external;0.8em`](https://mdsanima-dev.github.io/mdsanima-dev/)
be sure to check it.

## [0.2.0](https://github.com/mdsanima-dev/mdsanima-dev/compare/v0.1.1...v0.2.0)

_Date:_ _2023-05-12_

{octicon}`alert;1em;sd-text-warning`{bdg-danger-line}`BREAKING CHANGES`

- dynamic version and drop python3.7 python3.8 and python3.9

{octicon}`file-code;1em;sd-text-secondary`{bdg-light-line}`REFACTOR`

- convert images to `webp` format closes [#49](https://github.com/mdsanima-dev/mdsanima-dev/issues/49) ([8e7a13c](https://github.com/mdsanima-dev/mdsanima-dev/commit/8e7a13c1a431d6b8025be0b566fee18ebce834b9))
- dynamic version and drop py3.7 py3.8 py3.9 closes [#48](https://github.com/mdsanima-dev/mdsanima-dev/issues/48) ([5555845](https://github.com/mdsanima-dev/mdsanima-dev/commit/5555845b8d22eee3cdaf97defe7195ade20d1310))
- moved and renamed ([cb0d226](https://github.com/mdsanima-dev/mdsanima-dev/commit/cb0d22622007d02b4926f72e37acb394eff84e7e))

{octicon}`pulse;1em;sd-text-secondary`{bdg-warning-line}`BUG FIXES`

- add `CODEOWNERS` file and adding author ([0e5bfe3](https://github.com/mdsanima-dev/mdsanima-dev/commit/0e5bfe32065358a96a125360b55c68b6c2ed0da9))
- dynamic version from `setuptools_scm` package ([b0d197c](https://github.com/mdsanima-dev/mdsanima-dev/commit/b0d197c517463af1164719a52cab1fd553482aa1))

{octicon}`server;1em;sd-text-secondary`{bdg-primary-line}`BUILD SYSTEM`

- **deps-dev:** bump wheel from 0.37.1 to 0.38.1 in /docs ([ca55d8a](https://github.com/mdsanima-dev/mdsanima-dev/commit/ca55d8aff621bd8d32b0ddada14ee80a31a83d2d))
- dev dependencies ([8f46855](https://github.com/mdsanima-dev/mdsanima-dev/commit/8f46855987e7e11c9e73cecb75d9895770e85e34))
- keep this direcotry ([399ffd0](https://github.com/mdsanima-dev/mdsanima-dev/commit/399ffd04e29b89fa9c75693b969a578f061c44de))
- new config flake8 pylint and max line length closes [#47](https://github.com/mdsanima-dev/mdsanima-dev/issues/47) ([76fc8b6](https://github.com/mdsanima-dev/mdsanima-dev/commit/76fc8b690c24f5e8a74e08242bbc969c16013c2e))
- new config setup ([2bdc17c](https://github.com/mdsanima-dev/mdsanima-dev/commit/2bdc17c2223eedab2fab5e3ddb029d98594a3333))
- pylint ignore ([6f7abb3](https://github.com/mdsanima-dev/mdsanima-dev/commit/6f7abb3bcd7dec5d94760445c7999bb49d411e15))
- sort and remove unnecessary package ([48d9784](https://github.com/mdsanima-dev/mdsanima-dev/commit/48d97840ccdf02a127fada3001676c4fbd3c1d4b))
- update year ([4167e9d](https://github.com/mdsanima-dev/mdsanima-dev/commit/4167e9d987e9c994ded229f4a7f230567d32e665))

{octicon}`book;1em;sd-text-secondary`{bdg-secondary-line}`DOCUMENTATION`

- add image to getting started page ([6f70059](https://github.com/mdsanima-dev/mdsanima-dev/commit/6f70059ee9a709bea84599c9a773c93b994789e0))
- add info for change default branch to `main` on remote repo ([fac5c48](https://github.com/mdsanima-dev/mdsanima-dev/commit/fac5c486030d60a6194809ebfc7617879fdd26c3))
- remove instalation instruction using `setup` and `whell` ([3f78f3a](https://github.com/mdsanima-dev/mdsanima-dev/commit/3f78f3adce604af924d66931706d6ee339674273))

## [0.1.1](https://github.com/mdsanima-dev/mdsanima-dev/compare/v0.1.0...v0.1.1)

_Date:_ _2022-05-02_

{octicon}`pulse;1em;sd-text-secondary`{bdg-warning-line}`BUG FIXES`

- added email address ([bd600ab](https://github.com/mdsanima-dev/mdsanima-dev/commit/bd600ab56c73d70b81e5f8631d512df55f6eaae1))
- **build-system:** add new keywords and classifiers ([90d4cc9](https://github.com/mdsanima-dev/mdsanima-dev/commit/90d4cc9eacfe98ba70bf4115c0f11a8ab2c27e55))
- **build-system:** black format change variable name ([ef5f0b9](https://github.com/mdsanima-dev/mdsanima-dev/commit/ef5f0b9dfd746e077a07126f595078eda145c29a))
- **build-system:** configuration for black python code formatter ([a8100e1](https://github.com/mdsanima-dev/mdsanima-dev/commit/a8100e156eb8e9cbe15e21ce07b946cc59781dfb))
- **build-system:** isort options for sorting import ([4f93078](https://github.com/mdsanima-dev/mdsanima-dev/commit/4f930782b4e19bce1e6af73046ea9d06ef63f3d5))
- **build-system:** sidebar name change brand text size ([c9e1e4f](https://github.com/mdsanima-dev/mdsanima-dev/commit/c9e1e4fbb94731b86fd85377d5cced48ab2ab624))
- **console-scripts:** fix shell script `mdsanima-dev-converts` name ([b10e65f](https://github.com/mdsanima-dev/mdsanima-dev/commit/b10e65fcf4e6565974652b6ed905c2c2f4fda4bf))
- **devs:** black profile formater ([4cfca93](https://github.com/mdsanima-dev/mdsanima-dev/commit/4cfca9320cb0a872f4cfab7b68cb3597b5031e45))
- add json config on `.editorconfig` ([745107c](https://github.com/mdsanima-dev/mdsanima-dev/commit/745107cccb892a8dca3009c25a0b0a6c3b1c01ce))
- added current year to `LICENSE` file ([69eb229](https://github.com/mdsanima-dev/mdsanima-dev/commit/69eb2290dfaa8ca326135be04f6bd94a1a5028bb))

{octicon}`tools;1em;sd-text-secondary`{bdg-danger-line}`TESTS`

- print pacage data from json ([0db5e2e](https://github.com/mdsanima-dev/mdsanima-dev/commit/0db5e2e7862dbb8720d7a877c15b7115eab65d40))
- rename file and fix variable the same on `setyp.py` file ([b16a061](https://github.com/mdsanima-dev/mdsanima-dev/commit/b16a0615af82de90f122983673070a804c96a579))
- variable the same on `setyp.py` file ([f52a40d](https://github.com/mdsanima-dev/mdsanima-dev/commit/f52a40d5fec9f76670e838df9bf994817d98a5b3))


src/mdsanima_dev/colors.py
# Copyritht © 2022 Marcin Różewski MDSANIMA


"""Module :bdg-success-line:`colors` is for printing complex colors in the
console output with your own texts. This module is for also for showing example
colors output in the shell terminal window. You can use this output on your
future projects.

The core functionality is for replace default :bdg-primary-line:`print`
function to allow printing your string in colors on the terminal window. You
can use many default colors available on ``int`` values or you can use
a default color names scheme like ``blue`` or ``red`` options.
"""


from __future__ import annotations

import os
from typing import Union


INFO_KEY_ERROR = "Please choose a name from this default colors scheme"
INFO_OPT_ERROR = "Please choose from this 'number', 'name' or 'text' options"
DEFAULT_NAMES_NORMAL = {
    "cyan": 50,
    "sky": 39,
    "blue": 27,
    "indigo": 105,
    "violet": 99,
    "purple": 170,
    "fuchsia": 177,
    "pink": 206,
    "rose": 203,
    "red": 196,
    "orange": 208,
    "amber": 215,
    "yellow": 220,
    "lime": 148,
    "green": 118,
    "emerald": 122,
    "teal": 43,
    "graphe": 24,
    "giter": 30,
    "slate": 250,
    "gray": 245,
    "zinc": 238,
    "white": 255,
    "black": 16,
}


def _get_normal_default(color_name: str = "white") -> int:
    """Define default color scheme with several name variants for normal mode
    option that getting color number.

    .. warning:: Do not use this function in your code, it is needed to perform
        other functions.

    :param color_name: Choose a color name string instead of typing number,
        defaults to ``white``.
    :type color_name: str, optional
    :return: Color number for text foreground and background syntax output.
        Return type ``int``.
    """
    color_number = DEFAULT_NAMES_NORMAL[color_name]
    return color_number


def _set_normal_syntax() -> str:
    """Define colors variables foreground and background ``x1b`` syntax with
    normal mode for printing text with 256 colors.

    .. warning:: Do not use this function in your code, it is needed to perform
        other functions.

    :return: Color console literal variable for text foreground and background.
        Return type ``str``.
    """
    fg = "\x1b[38;5;"
    bg = "\x1b[48;5;"
    xm = "m"
    ex = "\x1b[0m"
    return fg, bg, xm, ex


def use_normal_mode(
    text: str = "mdsanima",
    fg_color: Union[str, int] = 255,
    bg_color: Union[str, int] = None,
) -> str:
    """This feature returning string with ``x1b`` syntax that allow to print
    the text with 256 colors normal mode on your output.

    :param text: The text foreground that's you want to print with colors,
        defaults to ``mdsanima``.
    :type text: str, optional
    :param fg_color: The color number or default name for text foreground.
        Choose a number from 0 to 255 or use color name, defaults to ``255``.
    :type fg_color: Union[str, int], optional
    :param bg_color: The color number or default name for background.
        Choose a number from 0 to 255 or use color name, defaults to ``None``.
    :type bg_color: Union[str, int], optional
    :return: Text string output with colors ``x1b`` systax.
        Return type ``str``.
    """
    fg, bg, xm, ex = _set_normal_syntax()
    try:
        if type(fg_color) == str:
            fg_color = _get_normal_default(fg_color)
        if type(bg_color) == str:
            bg_color = _get_normal_default(bg_color)
        if bg_color == None:
            foreground = fg + str(fg_color) + xm
            out_result = foreground + text + ex
        else:
            foreground = fg + str(fg_color) + xm
            background = bg + str(bg_color) + xm
            out_result = foreground + background + text + ex
        return out_result
    except KeyError:
        print(INFO_KEY_ERROR, DEFAULT_NAMES_NORMAL.keys())


def print_normal_mode(
    text: str = "mdsanima",
    fg_color: Union[str, int] = 255,
    bg_color: Union[str, int] = None,
    end=None,
) -> None:
    """This feature allows to print text and background with ``x1b`` syntax
    256 colors normal mode on your output. The function works the same like
    standard print function.

    :param text: The text foreground that's you want to print with colors,
        defaults to ``mdsanima``.
    :type text: str, optional
    :param fg_color: The color number or default name for text foreground.
        Choose a number from 0 to 255 or use color name, defaults to ``255``.
    :type fg_color: Union[str, int], optional
    :param bg_color: The color number or default name for background.
        Choose a number from 0 to 255 or use color name, defaults to ``None``.
    :type bg_color: Union[str, int], optional
    :param end: End of line print options, defaults to ``None``.
    :type end: None, optional
    :return: Printing colored string text foreground and background output.
        Return type ``None``.
    """
    out_result = use_normal_mode(text, fg_color, bg_color)
    return print(out_result, end=end)


def show_normal_colors(options: str = "number") -> str:
    """This function is for showing all available colors number or string
    example text output.

    :param options: What example will be show on print. Choose from this
        ``number``, ``name`` or ``text`` options, defaults to ``number``.
    :type options: str, optional
    :return: Printing all colors number, name or text example output.
        Return type ``str``.
    """
    fg, bg, xm, ex = _set_normal_syntax()

    # Setup colors for info print.
    colors = [196, 208, 70, 180, 0]

    # Setup unicode character.
    unicode_one = "\U0000005C"  # \
    unicode_two = "\U0000002F"  # /

    # Get terminal size columns for center info print.
    terminal = os.get_terminal_size()

    # Calculation terminal size based on 22 count string from info variables.
    calculate_befor = terminal.columns / 2 - 11
    calculate_after = terminal.columns / 2 - 11

    # Print across terminal size.
    befor = unicode_one * int(calculate_befor)
    after = unicode_two * int(calculate_after)

    # Setup colors info variables.
    cts = fg + str(colors[4]) + xm + bg + str(colors[0]) + xm + befor + ex
    cna = fg + str(colors[0]) + xm + "[" + ex
    csh = fg + str(colors[1]) + xm + " SHOW COLORS " + ex
    cdo = fg + str(colors[2]) + xm + " DONE COLORS " + ex
    cnu = fg + str(colors[3]) + xm + "NUMBER " + ex
    cnn = fg + str(colors[3]) + xm + " NAMES " + ex
    cmd = fg + str(colors[3]) + xm + " TEXTS " + ex
    cnc = fg + str(colors[0]) + xm + "]" + ex
    cte = fg + str(colors[4]) + xm + bg + str(colors[0]) + xm + after + ex

    # Print info colors options.
    if options == "number":
        print(cts + cna + csh + cnu + cnc + cte)
        done = "\n" + cte + cna + cdo + cnu + cnc + cts
    if options == "name":
        print(cts + cna + csh + cnn + cnc + cte)
        done = "\n" + cte + cna + cdo + cnn + cnc + cts
    if options == "text":
        print(cts + cna + csh + cmd + cnc + cte)
        done = "\n" + cte + cna + cdo + cmd + cnc + cts
    else:
        done = INFO_OPT_ERROR

    # Print all colors numbers.
    if options == "number" or options == "text":
        for i in range(256):
            if options == "number":
                show = "--> " + str(i).ljust(5)
            if options == "text":
                show = "mdsanima".ljust(9)
            color = fg + str(i) + xm + show + ex
            print(color, sep=" ", end="", flush=True)
    if options == "name":
        default_key = DEFAULT_NAMES_NORMAL.keys()
        for name in list(default_key):
            show = str(name) + " "
            color = fg + str(DEFAULT_NAMES_NORMAL[name]) + xm + show + ex
            print(color, sep=" ", end="", flush=True)

    return print(done)
